### PR TITLE
feat(compiler): variable usage is allowed

### LINKED ISSUE #312: validation: variables
- [x] variables in operation definitions must be unique, [spec](https://spec.graphql.org/October2021/#sec-Variable-Uniqueness)
- [x] variables must be of input type, [spec](https://spec.graphql.org/October2021/#sec-Variables-Are-Input-Types)
- [ ] all varaible usages are allowed, [spec](https://spec.graphql.org/October2021/#sec-All-Variable-Usages-are-Allowed)

### PR DESCRIPTION
**Pull request by lrlna**
closes #312 

This looks at currently defined variables and compares them to how they are used, [spec](https://spec.graphql.org/October2021/#sec-All-Variable-Usages-are-Allowed).

### CODE REVIEW

File: crates/apollo-compiler/src/validation/fragment.rs [hunk 1]
```diff
@@ -180,54 +183,60 @@ pub fn validate_inline_fragment(
...
-        diagnostics.extend(db.validate_fragment_used(Arc::clone(def), file_id));
+    diagnostics.extend(db.validate_directives(
+        def.directives().to_vec(),
+        hir::DirectiveLocation::FragmentDefinition,
+        var_defs.clone(),
+    ));
+    diagnostics.extend(
+        db.validate_fragment_type_condition(Some(def.type_condition().to_string()), def.loc()),
+    );
+    diagnostics.extend(db.validate_selection_set(def.selection_set().clone(), var_defs));
```
Comment (lrlna) [inline crates/apollo-compiler/src/validation/fragment.rs:237]: The consequence of moving `db.validate_fragment_definition` to be run during selection validation is that we somehow end up missing some field validation inside a fragment definition's selection set! really confused on why that's happening

File: crates/apollo-compiler/test_data/diagnostics/0089_fragment_type_condition_on_non_existent_types.txt [hunk 2]
```diff
@@ -39,30 +39,61 @@
...
+                text: "`a` field is not defined",
+            },
+        ],
+        help: Some(
+            "`a` is not defined on `Interface2` type",
+        ),
+        data: UndefinedField {
+            field: "a",
+        },
+    },
```
Comment (lrlna) [inline crates/apollo-compiler/test_data/diagnostics/0089_fragment_type_condition_on_non_existent_types.txt:96]: I feel like this diagnostic here should not be here! something went astray with fragment validation 😬

File: crates/apollo-compiler/test_data/diagnostics/0090_fragment_spread_impossible.txt
Comment (lrlna) [inline crates/apollo-compiler/test_data/diagnostics/0090_fragment_spread_impossible.txt:1]: something went wrong with validation here also, as these diagnostics should definitely be here!

File: crates/apollo-compiler/test_data/diagnostics/0088_fragment_selection_set.graphql [hunk 3]
```diff
@@ -13,6 +13,9 @@ query getProduct {
     topProduct {
         ...productFragment
     }
+    topProducts {
+        ...productFragment
+    }
```
Comment (lrlna) [inline crates/apollo-compiler/test_data/diagnostics/0088_fragment_selection_set.graphql:18]: the fragment won't try to be validated if we don't call it, so calling it again inside a correct field

File: crates/apollo-compiler/test_data/diagnostics/0090_fragment_spread_impossible.graphql
Comment (lrlna) [inline crates/apollo-compiler/test_data/diagnostics/0090_fragment_spread_impossible.graphql:1]: @goto-bus-stop can you triple check that validations for this test are correct? i had to add a query to call these fragments, but am not sure if that produces the right diagnostics

File: crates/apollo-compiler/src/validation/fragment.rs (hunk 1 above)
Comment (lrlna) [reply crates/apollo-compiler/src/validation/fragment.rs]: fixed this bit!

File: crates/apollo-compiler/test_data/diagnostics/0089_fragment_type_condition_on_non_existent_types.txt (hunk 2 above)
Comment (lrlna) [reply crates/apollo-compiler/test_data/diagnostics/0089_fragment_type_condition_on_non_existent_types.txt]: fixed this also!

File: crates/apollo-compiler/test_data/diagnostics/0090_fragment_spread_impossible.txt
Comment (lrlna) [reply crates/apollo-compiler/test_data/diagnostics/0090_fragment_spread_impossible.txt:1]: also fixed

File: crates/apollo-compiler/test_data/diagnostics/0006_subscription_with_multiple_root_fields_in_inline_fragments.txt [hunk 4]
```diff
@@ -1,4 +1,45 @@
...
+                text: "`Subscription` declared here",
+            },
+        ],
+        help: Some(
+            "`disallowedSecondRootField` is not defined on `Subscription` type",
+        ),
+        data: UndefinedField {
+            field: "disallowedSecondRootField",
+        },
+    },
```
Comment (goto-bus-stop) [inline crates/apollo-compiler/test_data/diagnostics/0006_subscription_with_multiple_root_fields_in_inline_fragments.txt:42]: to keep this test focused on just the multiple root fields diagnostic, maybe we should define `disallowedSecondRootField` in `type Subscription`?

**Review by goto-bus-stop — CHANGES_REQUESTED** [review]
Looks great! \ :grin: /

File: crates/apollo-compiler/src/validation/variable.rs [hunk 5]
```diff
@@ -137,6 +126,161 @@ pub fn validate_unused_variables(
...
+            }
+            hir::Type::List { ty: loc_ty, .. } => return are_types_compatible(loc_ty, variable_ty),
+            hir::Type::Named { .. } => return are_types_compatible(location_ty, variable_ty),
+        }
+    }
+
+    are_types_compatible(location_ty, variable_ty)
+}
+
+fn are_types_compatible(location_ty: &hir::Type, variable_ty: &hir::Type) -> bool {
```
[the remark this answers is not in the document]
Comment (lrlna) [reply crates/apollo-compiler/src/validation/variable.rs:231]: ah good catch, yes let's

File: crates/apollo-compiler/src/validation/variable.rs [hunk 6]
```diff
@@ -137,6 +126,161 @@ pub fn validate_unused_variables(
...
+            hir::Type::List {
+                ty: item_variable_ty,
+                ..
+            } => return are_types_compatible(item_location_ty, item_variable_ty),
+            // 3.a. If variable_ty is NOT a list type, return false.
+            hir::Type::Named { .. } => return false,
+            // 3.a. If variable_ty is NOT a list type, return false.
+            hir::Type::NonNull { .. } => return false,
+        },
+    };
```
[the remark this answers is not in the document]
Comment (lrlna) [reply crates/apollo-compiler/src/validation/variable.rs:280]: yea!

**Pull request merged**